grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java

```diff
@@ -209,66 +208,78 @@ protected final void maybeCancel() {
     }
 
     public final void close(Throwable exception) {
+        close(exception, false);
+    }
+
+    public final void close(Throwable exception, boolean cancelled) {
         exception = Exceptions.peel(exception);
         final Metadata metadata = generateMetadataFromThrowable(exception);
-        close(GrpcStatus.fromThrowable(exceptionHandler, ctx, exception, metadata), metadata, exception);
+        final Status status = GrpcStatus.fromThrowable(exceptionHandler, ctx, exception, metadata);
+        close(new ServerStatusAndMetadata(status, metadata, cancelled), exception);
     }
 
     @Override
     public final void close(Status status, Metadata metadata) {
-        close(GrpcStatus.fromExceptionHandler(exceptionHandler, ctx, status, metadata), metadata, null);
+        final ServerStatusAndMetadata statusAndMetadata =
+                new ServerStatusAndMetadata(GrpcStatus.fromExceptionHandler(exceptionHandler, ctx,
+                                                                            status, metadata),
+                                            metadata, false);
+        close(statusAndMetadata, null);
     }
 
-    private void close(Status status, Metadata metadata, @Nullable Throwable exception) {
+    private void close(ServerStatusAndMetadata statusAndMetadata, @Nullable Throwable exception) {
         if (ctx.eventLoop().inEventLoop()) {
-            doClose(status, metadata, exception);
+            doClose(statusAndMetadata, exception);
         } else {
             ctx.eventLoop().execute(() -> {
-                doClose(status, metadata, exception);
+                doClose(statusAndMetadata, exception);
             });
         }
     }
 
-    private void doClose(Status status, Metadata metadata, @Nullable Throwable exception) {
+    private void doClose(ServerStatusAndMetadata statusAndMetadata, @Nullable Throwable exception) {
         maybeLogFailedRequestContent(exception);
+        Status status = statusAndMetadata.status();
+        final Metadata metadata = statusAndMetadata.metadata();
         if (isCancelled()) {
             // No need to write anything to client if cancelled already.
-            closeListener(status, metadata, false, true);
+            statusAndMetadata.completed(false);
+            statusAndMetadata.setResponseContent(true);
+            closeListener(statusAndMetadata);
             return;
         }
 
         if (status.getCode() == Code.CANCELLED && status.getCause() instanceof ClosedStreamException) {
```

Comment:
`ServerStatusAndMetadata.cancelled` could be set to true when it is created and then, we may unify the two conditions.
```java
if (serverStatusAndMetadata.isCancelled()) {
   ...
}
```